fix(ingest): support `database_pattern` in clickhouse and clickhouse-usage

Gerrit-K opened the pull request:
This changes the `clickhouse` and `clickhouse-usage` ingestion sources to respect the config value `database_pattern` instead of the deprecated `schema_pattern`. The latter can't be used anymore since its deprecation, so it's currently impossible to filter for databases in these modules.

Note: there are also a lot of other occurrences of "schema" in the code that might be renamed to "database" as well, but I didn't want to do a full blown refactoring here. If you think it is useful, let me know and I'll adjust the PR.


## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

hsheth2 commented:
@mayurinehate could you take a look at this PR when you get a chance? It seems like two-tier SQL sources should not have both database_pattern and schema_pattern. It probably makes sense to hide/remove `schema_pattern` from the docs/configs.

Gerrit-K commented:
Note: after submitting this PR, I also noticed that the table pattern doesn't work here as expected. Based on the docs and the other sources, it should filter on the full table name including the database name. But here, it just filters on the table name alone (which makes it impossible to target a particular table in a specific database when there is a similarly named table in another database). I unfortunately didn't find the time yet to update the PR accordingly.

Gerrit-K commented:
> Note: after submitting this PR, I also noticed that the table pattern doesn't work here as expected. Based on the docs and the other sources, it should filter on the full table name including the database name. But here, it just filters on the table name alone (which makes it impossible to target a particular table in a specific database when there is a similarly named table in another database). I unfortunately didn't find the time yet to update the PR accordingly.

I updated my PR accordingly:
- `table_pattern` now filters on the full dataset name (including the database) for both, `clickhouse` and `clickhouse-usage`
- Dropped lineage events are included in the report
- Dropped usage events are logged (because the report doesn't have a `report_dropped` method here)

Happy to hear feedback on this!

mayurinehate reviewed (approved):
lgtm

mayurinehate commented:
> @mayurinehate could you take a look at this PR when you get a chance? It seems like two-tier SQL sources should not have both database_pattern and schema_pattern. It probably makes sense to hide/remove `schema_pattern` from the docs/configs.

Yeah, this is due to config hierarchy. For 3-tier sql sources, both database_pattern and schema_pattern make sense .

@Gerrit-K could you add `hide_from_schema=True,` in `schema_pattern` config in [two_tier_sql_source](https://github.com/Gerrit-K/datahub/blob/fix/clickhouse-ingestion-database-pattern/metadata-ingestion/src/datahub/ingestion/source/sql/two_tier_sql_source.py#L33) config ?

Gerrit-K commented:
> @Gerrit-K could you add `hide_from_schema=True,` in `schema_pattern` config in [two_tier_sql_source](https://github.com/Gerrit-K/datahub/blob/fix/clickhouse-ingestion-database-pattern/metadata-ingestion/src/datahub/ingestion/source/sql/two_tier_sql_source.py#L33) config ?

Like so? https://github.com/datahub-project/datahub/pull/9482/commits/c6f257e9e22e76d9dfe97ee5b2e923c2644c432b

mayurinehate commented:
> > @Gerrit-K could you add `hide_from_schema=True,` in `schema_pattern` config in [two_tier_sql_source](https://github.com/Gerrit-K/datahub/blob/fix/clickhouse-ingestion-database-pattern/metadata-ingestion/src/datahub/ingestion/source/sql/two_tier_sql_source.py#L33) config ?
> 
> Like so? [c6f257e](https://github.com/datahub-project/datahub/commit/c6f257e9e22e76d9dfe97ee5b2e923c2644c432b)

Yeah, looks like it didn't work. Probably its `hidden_from_docs` and not `hide_from_schema`. Could you make this change ?

Gerrit-K commented:
> Yeah, looks like it didn't work. Probably its `hidden_from_docs` and not `hide_from_schema`. Could you make this change ?

Applied it 👍 Not sure what exactly to look out for, but if we're expecting the `schema_pattern` to vanish from the doc pages, it seems like it still didn't work 🤔

mayurinehate commented:
It has worked - https://docs-website-djmtxduoz-acryldata.vercel.app/docs/next/generated/ingestion/sources/clickhouse

Gerrit-K commented:
> It has worked - https://docs-website-djmtxduoz-acryldata.vercel.app/docs/next/generated/ingestion/sources/clickhouse
> 
> It'll reflect on actual docs page after the PR is merged.

Oh, okay, maybe the environment I was looking at was not updated yet when I checked :)

Thanks for taking a look, btw!

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/usage/clickhouse_usage.py:
```diff
@@ -150,9 +151,12 @@ def _get_clickhouse_history(self):
-            ) or not self.config.table_pattern.allowed(event_dict.get("table")):
+            if (
+                    not self.config.database_pattern.allowed(event_dict.get("database")) or
+                    not self.config.table_pattern.allowed(event_dict.get("full_table_name")) or
+                    not self.config.view_pattern.allowed(event_dict.get("full_table_name"))
```
we only want to skip if it is not allowed by both the table pattern and the view pattern
```suggestion
                    not (self.config.table_pattern.allowed(event_dict.get("full_table_name")) or
                   self.config.view_pattern.allowed(event_dict.get("full_table_name")))
```

hsheth2 reviewed (commented):
mostly looks good, one question about the usage filtering logic

hsheth2 commented:
@Gerrit-K the linter probably won't like the formatting on that if statement

Gerrit-K commented:
> @Gerrit-K the linter probably won't like the formatting on that if statement

Sorry, I was in a hurry and still haven't set up all the local tooling 🙈 hope this is better now

Gerrit-K commented:
@hsheth2 the CI failure seems unrelated to my changes (I hope). Do you happen to know if someone is already taking care of that?

hsheth2 commented:
@Gerrit-K yup that flaky test was fixed in a separate PR, so should be good to merge! Thanks for the contribution here :)
